client/peerset/src/protocol_controller.rs

```diff
@@ -0,0 +1,559 @@
+				}
+
+				if self.num_in >= self.max_in {
+					self.reject_connection(incoming_index);
+					return
+				}
+
+				self.num_in += 1;
+				*state = PeerState::Connected(Direction::Inbound);
+				self.accept_connection(incoming_index);
+			},
+		}
+	}
+
+	/// Returns Err(PeerId) if peer wasn't connected.
+	fn on_peer_dropped(&mut self, peer_id: PeerId, reason: DropReason) -> Result<(), PeerId> {
+		let mut is_reserved = true;
+		let state = self
+			.reserved_nodes
+			.get_mut(&peer_id)
+			.or_else(|| {
+				is_reserved = false;
+				self.nodes.get_mut(&peer_id)
+			})
+			.ok_or(peer_id)?;
+		match state {
+			PeerState::Connected(d) => {
+				if is_reserved {
+					*state = PeerState::NotConnected;
+				} else {
+					match d {
+						Direction::Inbound => self.num_in -= 1,
+						Direction::Outbound => self.num_out -= 1,
+					}
+					if let DropReason::Refused = reason {
+						self.nodes.remove(&peer_id);
+					} else {
+						*state = PeerState::NotConnected;
+					}
+				}
+				self.report_disconnect(peer_id);
+				Ok(())
+			},
+			PeerState::NotConnected => Err(peer_id),
+		}
+	}
+
+	fn alloc_slots(&mut self) {
+		if self.num_out >= self.max_out {
+			return
+		}
+
+		// Try connecting to reserved nodes first.
+		self.reserved_nodes
+			.iter_mut()
+			.filter_map(|(peer_id, state)| {
+				(matches!(state, PeerState::NotConnected) &&
+					!self.peerset_handle.is_banned(*peer_id))
+				.then_some({
+					*state = PeerState::Connected(Direction::Outbound);
+					peer_id
+				})
+			})
+			.cloned()
+			.collect::<Vec<_>>()
+			.into_iter()
+			.map(|peer_id| {
+				self.start_connection(peer_id);
+			});
+
+		// Nothing more to do if we're in reserved-only mode.
+		if self.reserved_only {
+			return
+		}
+
+		// Fill available slots with non-reserved nodes
+		let available_slots = (self.max_out - self.num_out).saturated_into();
+		// Ignore reserved nodes (connected above) and already connected nodes.
+		let ignored = self
+			.reserved_nodes
+			.keys()
+			.cloned()
+			.collect::<HashSet<PeerId>>()
+			.union(
+				&self
+					.nodes
+					.iter()
+					.filter_map(|(peer_id, state)| state.is_connected().then_some(*peer_id))
+					.collect::<HashSet<_>>(),
+			)
+			.cloned()
+			.collect::<HashSet<PeerId>>();
+
+		self.peerset_handle
+			.outgoing_candidates(available_slots, ignored)
+			.filter_map(|peer_id| {
+				let state = self.nodes.entry(*peer_id).or_insert(PeerState::NotConnected);
+				match state {
+					PeerState::Connected(_) => {
+						debug_assert!(false, "`Peerset` returned a node we asked to ignore.");
+						error!(
+							target: "peerset",
+							"`Peerset` returned a node we asked to ignore: {}.",
+							peer_id
+						);
+						None
+					},
+					PeerState::NotConnected => {
+						self.num_out += 1;
+						*state = PeerState::Connected(Direction::Outbound);
+						Some(peer_id)
+					},
+				}
+			})
+			.cloned()
+			.collect::<Vec<_>>()
+			.iter()
+			.map(|peer_id| self.start_connection(*peer_id));
```

Comment:
same here, `for_each()` is probably the right function to call

Reply:
Done